File: src/java/magmac/api/iter/Iter.java
package magmac.api.iter;

import magmac.api.iter.collect.Collector;
import magmac.api.result.Result;

import magmac.api.Option;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;

public interface Iter<T> {
    <R, X> Result<R, X> foldToResult(R initial, BiFunction<R, T, Result<R, X>> folder);

    <R> Iter<R> map(Function<T, R> mapper);

    <R> R fold(R initial, BiFunction<R, T, R> folder);

    <C> C collect(Collector<T, C> collector);

    Iter<T> filter(Predicate<T> predicate);

    Option<T> next();

    <R> Iter<R> flatMap(Function<T, Iter<R>> mapper);

    Iter<T> concat(Iter<T> other);
}

File: src/web/magmac/api/iter/Iter.ts
export interface Iter<T> {
	 foldToResult( initial : R,  folder : BiFunction<R, T, Result<R, X>>) : Result<R, X>;
	 map( mapper : Function<T, R>) : Iter<R>;
	 fold( initial : R,  folder : BiFunction<R, T, R>) : R;
	 collect( collector : Collector<T, C>) : C;
	 filter( predicate : Predicate<T>) : Iter<T>;
	 next() : Option<T>;
	 flatMap( mapper : Function<T, Iter<R>>) : Iter<R>;
	 concat( other : Iter<T>) : Iter<T>;
}
